File: WineCellar/Services/CSVImporter.swift
import Foundation

struct CSVRow {
    let values: [String]

    func value(at index: Int) -> String {
        guard index >= 0, index < values.count else { return "" }
        return values[index].trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

struct CSVData {
    let headers: [String]
    let rows: [CSVRow]
}

enum CSVImporter {
    static func parse(_ content: String) -> CSVData {
        let lines = content.components(separatedBy: .newlines).filter { !$0.trimmingCharacters(in: .whitespaces).isEmpty }
        guard let headerLine = lines.first else {
            return CSVData(headers: [], rows: [])
        }

        // Detect delimiter: tab or comma
        let delimiter: Character = headerLine.contains("\t") ? "\t" : ","

        let headers = parseLine(headerLine, delimiter: delimiter)
        let rows = lines.dropFirst().map { CSVRow(values: parseLine($0, delimiter: delimiter)) }

        return CSVData(headers: headers, rows: rows)
    }

    private static func parseLine(_ line: String, delimiter: Character = ",") -> [String] {
        var fields: [String] = []
        var current = ""
        var inQuotes = false

        for char in line {
            if char == "\"" {
                inQuotes.toggle()
            } else if char == delimiter && !inQuotes {
                fields.append(current)
                current = ""
            } else {
                current.append(char)
            }
        }
        fields.append(current)

        return fields.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
    }
}

struct ColumnMapping {
    var name: Int?
    var producer: Int?
    var variety: Int?
    var region: Int?
    var country: Int?
    var vintage: Int?
    var zone: Int?
    var slot: Int?
    var quantity: Int?
    var notes: Int?

    var mappedCount: Int {
        [name, producer, variety, region, country, vintage, zone, slot, quantity, notes].compactMap { $0 }.count
    }

    func buildWine(from row: CSVRow) -> Wine {
        Wine(
            name: name.map { row.value(at: $0) } ?? "",
            producer: producer.map { row.value(at: $0) } ?? "",
            variety: variety.map { row.value(at: $0) } ?? "",
            region: region.map { row.value(at: $0) } ?? "",
            country: country.map { row.value(at: $0) } ?? "",
            vintage: vintage.flatMap { Int(row.value(at: $0)) } ?? Calendar.current.component(.year, from: Date()),
            zone: zone.map { row.value(at: $0) } ?? "",
            slot: slot.flatMap { Int(row.value(at: $0)) } ?? 1,
            notes: notes.map { row.value(at: $0) } ?? "",
            quantity: quantity.flatMap { Int(row.value(at: $0)) } ?? 1
        )
    }
}

File: wine-cellar-pwa/app/lib/csv/csvImporter.ts
import type { Wine } from '@/app/types'

interface ColumnMapping {
  name?: number
  producer?: number
  variety?: number
  region?: number
  country?: number
  vintage?: number
  zone?: number
  notes?: number
  drankIdx?: number
}

function parseLine(line: string, delimiter: string): string[] {
  const fields: string[] = []
  let current = ''
  let inQuotes = false
  for (const char of line) {
    if (char === '"') {
      inQuotes = !inQuotes
    } else if (char === delimiter && !inQuotes) {
      fields.push(current.trim())
      current = ''
    } else {
      current += char
    }
  }
  fields.push(current.trim())
  return fields
}

function autoMap(headers: string[]): ColumnMapping {
  const mapping: ColumnMapping = {}
  const normalized = headers.map(h =>
    h.replace(/\uFEFF/g, '').toLowerCase().trim()
  )

  for (let i = 0; i < normalized.length; i++) {
    const h = normalized[i]
    if (h.includes('maker') || h.includes('producer') || h.includes('winery') || h.includes('brand')) {
      mapping.producer = i
    } else if (h === 'wine') {
      mapping.variety = i
    } else if (h.includes('name') && !h.includes('zone')) {
      mapping.name = i
    } else if (h.includes('variety') || h.includes('varietal') || h.includes('grape')) {
      mapping.variety = mapping.variety ?? i
    } else if (h.includes('country')) {
      mapping.country = i
    } else if (h.includes('region') || h.includes('appellation')) {
      mapping.region = i
    } else if (h.includes('vintage') || h.includes('year')) {
      mapping.vintage = i
    } else if (h.includes('zone') || h.includes('location')) {
      mapping.zone = i
    } else if (h.includes('note')) {
      mapping.notes = i
    } else if (h.includes('drank') || h.includes('drunk') || h.includes('consumed')) {
      mapping.drankIdx = i
    }
  }
  return mapping
}

export interface ParsedCSV {
  wines: Omit<Wine, 'id' | 'date_added'>[]
  skipped: number
  total: number
}

export function parseCSV(content: string, cellarId: string): ParsedCSV {
  const clean = content.replace(/\uFEFF/g, '')
  const lines = clean.split(/\r?\n/).filter(l => l.trim())
  if (lines.length < 2) return { wines: [], skipped: 0, total: 0 }

  const delimiter = lines[0].includes('\t') ? '\t' : ','
  const headers = parseLine(lines[0], delimiter)
  const mapping = autoMap(headers)

  const wines: Omit<Wine, 'id' | 'date_added'>[] = []
  let skipped = 0
  const currentYear = new Date().getFullYear()

  for (let i = 1; i < lines.length; i++) {
    const values = parseLine(lines[i], delimiter)
    const get = (idx?: number) => idx !== undefined ? (values[idx] ?? '').trim() : ''

    // Skip consumed wines
    if (mapping.drankIdx !== undefined) {
      const drank = get(mapping.drankIdx).toLowerCase()
      if (drank === 'true' || drank === 'yes' || drank === '1') {
        skipped++
        continue
      }
    }

    const name = get(mapping.name)
    const producer = get(mapping.producer)
    const variety = get(mapping.variety)

    // Skip empty rows
    if (!name && !variety && !producer) { skipped++; continue }

    const vintageStr = get(mapping.vintage)
    const vintage = parseInt(vintageStr) || 0
    const validVintage = vintage >= 1900 && vintage <= currentYear ? vintage : 0

    wines.push({
      cellar_id: cellarId,
      name,
      producer,
      variety,
      region: get(mapping.region),
      country: get(mapping.country),
      vintage: validVintage,
      zone: get(mapping.zone),
      slot: 1,
      notes: get(mapping.notes),
      quantity: 1,
      photo_path: null,
    })
  }

  return { wines, skipped, total: lines.length - 1 }
}
